sstables/sstable_set.cc

```diff
@@ -865,9 +873,15 @@ class incremental_reader_selector : public reader_selector {
         _pr = &pr;
 
         auto pos = dht::ring_position_view::for_range_start(*_pr);
+
         if (dht::ring_position_tri_compare(*_s, pos, _selector_position) >= 0) {
             return create_new_readers(pos);
         }
+        // If selector position Y is contained in new range [X, Z], then we should try selecting new
+        // sstables since it might have sstables that overlap with that range.
+        if (!_selector_position.is_max() && dht::ring_position_tri_compare(*_s, _selector_position, pr_end()) <= 0) {
+            return create_new_readers(std::nullopt);
```

Comment:
I don't understand why would this call return readers when the previous didn't.

Reply:
Consider scan of range [0, 10]

Position 5 is cached

sstables: [0, 5], [7, 10]

1) cache reader emits read on range [0, 5)
2) selects sstable [0, 5], next position: 7
3) doesn't consume sstable [7, 10] since it doesn't overlap with [0, 5)
4) cache reader fast forward to range [6, 10]

we wouldn't consume sstable [7, 10] without the change, because [6, 10] is not after next position of 7, but rather overlaps with it.

this change is required after we patched incremental reader selector to not select beyond the requested range (the EOS).

Reply:
the last case of test test_tablet_sstable_set_fast_forward_across_tablet_ranges catches it.

Reply:
Hmm, this sounds like it was a general problem with fast-forwarding in the selector, but we never caught it because with vnodes we never had a trigger condition.
Calling `create_new_readers()` with `std::nullopt` is like doing a cold start - it will just return the first sstables in the interval. It makes sense to do this always after fast-forward to? Do we still need the old if?

Reply:
the new range is [x, z]. the selector position is y, so naturally it's ok to call create_new_readers(nullopt), which is essentially the same as create_new_readers(y).

the old if makes sense. if new range is past selector position, then call create_new_readers(new_range_start).
the new if is essentially a overlap check with new_range, since once we're past the old if, new range start is before the selector position, then we also check if selector position is before the new range end.